File: lib-segment/src/main/java/websiteschema/mpsegment/core/POSRecognizer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.core;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import websiteschema.mpsegment.dict.IWord;
import websiteschema.mpsegment.dict.POSArray;
import websiteschema.mpsegment.graph.IGraph;
import websiteschema.mpsegment.graph.Path;
import websiteschema.mpsegment.hmm.*;
import websiteschema.mpsegment.util.SerializeHandler;

/**
 *
 * @author twer
 */
public class POSRecognizer implements IPOSRecognizer {

    private final Viterbi viterbi;
    private IGraph graph;
    private Path path;
    private int[] posFreq;
    private List<String> observeList = new ArrayList<String>();

    public POSRecognizer() {
        viterbi = new Viterbi();
        viterbi.setSortor(new TrieNodeQuickSort());
        load();
    }

    private void initViterbiWithEmissionAndObserve() {
        observeList.clear();
        viterbi.setE(new Emission());
        viterbi.setObserveBank(new NodeRepository());
        for (int i = 0; i < path.getLength(); i++) {
            IWord word = getWordIndexAt(i);
            initViterbiWith(word);
        }
    }

    private void initViterbiWith(IWord word) {
        String wordName = word.getWordName();
        Node observe = new Node(wordName);
        observeList.add(wordName);
        viterbi.getObserveBank().add(observe);
        POSArray posArray = word.getPOSArray();
        int[][] posTable = posArray.getWordPOSTable();
        for (int i = 0; i < posTable.length; i++) {
            int pos = posTable[i][0];
            int freq = posTable[i][1];
            viterbi.getE().setProb(pos, observe.getIndex(), getEmissionProb(pos, freq));
        }
    }

    private double getEmissionProb(int pos, int freq) {
        int tagFreq = getTagFreqs(pos);
        if (tagFreq > 0) {
            return (double) (freq + 1) / (double) tagFreq;
        } else {
            return 0.00001D;
        }
    }
    
    private int getTagFreqs(int i) {
        return posFreq[i];
    }

    private IWord getWordIndexAt(int index) {
        return graph.getEdgeObject(path.iget(index), path.iget(index + 1));
    }

    @Override
    public int[] findPOS(Path p, IGraph graph) {
        this.path = p;
        this.graph = graph;
        initViterbiWithEmissionAndObserve();
        try {
            List<Node> stateList = viterbi.calculateWithLog(observeList);
            int[] posArray = new int[stateList.size()];
            int i = 0;
            for (Node state : stateList) {
                posArray[i++] = state.getIndex();
            }
            return posArray;
        } catch (ObserveListException ex) {
            throw new RuntimeException("Could not find POS array." + ex.getMessage(), ex);
        }
    }

    public void save(SerializeHandler writeHandler) throws IOException {
        writeHandler.serializeArrayInt(posFreq);
        viterbi.getTran().save(writeHandler);
        viterbi.getPi().save(writeHandler);
        viterbi.getStateBank().save(writeHandler);
    }

    public final void load() {
        posFreq = POSResources.getInstance().getPosFreq();
        viterbi.setTran(POSResources.getInstance().getTransition());
        viterbi.setPi(POSResources.getInstance().getPi());
        viterbi.setStateBank(POSResources.getInstance().getStateBank());
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/core/POSRecognizer.scala
package websiteschema.mpsegment.core

import websiteschema.mpsegment.dict.IWord
import websiteschema.mpsegment.graph.IGraph
import websiteschema.mpsegment.graph.Path
import websiteschema.mpsegment.hmm._
import websiteschema.mpsegment.util.SerializeHandler
import collection.mutable.ListBuffer

class POSRecognizer extends IPOSRecognizer {

  private var viterbi: HmmViterbi = null
  private var graph: IGraph = null
  private var path: Path = null
  private var posFreq: Array[Int] = null
  private var observeList = ListBuffer[String]()

  viterbi = new HmmViterbi()
  Trie.setTreeNodeSorter(new TrieNodeQuickSort())
  load()

  private def initViterbiWithEmissionAndObserve() {
    observeList.clear()
    viterbi.setE(Emission())
    viterbi.setObserveBank(new NodeRepository())
    for (i <- 0 until path.getLength())
    {
      val word = getWordIndexAt(i)
      initViterbiWith(word)
    }
  }

  private def initViterbiWith(word: IWord) {
    val wordName = word.getWordName()
    val observe = Node(wordName)
    observeList += (wordName)
    viterbi.getObserveBank().add(observe)
    val posArray = word.getPOSArray()
    val posTable = posArray.getWordPOSTable()
    for (i <- 0 until posTable.length)
    {
      val pos = posTable(i)(0)
      val freq = posTable(i)(1)
      viterbi.getE().setProb(pos, observe.getIndex(), getEmissionProb(pos, freq))
    }
  }

  private def getEmissionProb(pos: Int, freq: Int): Double = {
    val tagFreq = getTagFreqs(pos)
    if (tagFreq > 0) {
      return (freq + 1).toDouble / tagFreq.toDouble
    } else {
      return 0.00001D
    }
  }

  private def getTagFreqs(i: Int): Int = {
    return posFreq(i)
  }

  private def getWordIndexAt(index: Int): IWord = {
    return graph.getEdgeObject(path.iget(index), path.iget(index + 1))
  }

  override def findPOS(p: Path, graph: IGraph): Array[Int] = {
    this.path = p
    this.graph = graph
    initViterbiWithEmissionAndObserve()
    try {
      val stateList = viterbi.calculateWithLog(observeList.toList)
      val posArray = new Array[Int](stateList.size)
      var i = 0
      for (state <- stateList) {
        posArray(i) = state.getIndex()
        i += 1
      }
      return posArray
    } catch {
      case ex: Throwable =>
      throw new RuntimeException("Could not find POS array." + ex.getMessage(), ex)
    }
  }

  def save(writeHandler: SerializeHandler) {
    POSResources.save(writeHandler)
  }

  def load() {
    posFreq = POSResources.getPosFreq()
    viterbi.setTran(POSResources.getTransition())
    viterbi.setPi(POSResources.getPi())
    viterbi.setStateBank(POSResources.getStateBank())
  }
}
